libraries/WiFi/src/WiFiGeneric.cpp

```diff
@@ -1450,35 +1448,82 @@ static void wifi_dns_found_callback(const char *name, const ip_addr_t *ipaddr, v
 }
 
 /**
- * Resolve the given hostname to an IP address. If passed hostname is an IP address, it will be parsed into IPAddress structure.
- * @param aHostname     Name to be resolved or string containing IP address
+ * IPv6 compatible DNS callback
+ * @param name
+ * @param ipaddr
+ * @param callback_arg
+ */
+static void wifi_dns6_found_callback(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    struct dns_api_msg *msg = (struct dns_api_msg *)callback_arg;
+
+    if(ipaddr && !msg->result) {
+        msg->ip_addr = *ipaddr;
+        msg->result = 1;
+    } else {
+        msg->result = -1;
+    }
+    xEventGroupSetBits(_arduino_event_group, WIFI_DNS_DONE_BIT);
+}
+
+/**
+ * Resolve the given hostname to an IP address.
+ * @param aHostname     Name to be resolved
  * @param aResult       IPAddress structure to store the returned IP address
  * @return 1 if aIPAddrString was successfully converted to an IP address,
  *          else error code
  */
 int WiFiGenericClass::hostByName(const char* aHostname, IPAddress& aResult)
 {
-    if (!aResult.fromString(aHostname))
-    {
-        ip_addr_t addr;
-        aResult = static_cast<uint32_t>(0);
-        waitStatusBits(WIFI_DNS_IDLE_BIT, 16000);
-        clearStatusBits(WIFI_DNS_IDLE_BIT | WIFI_DNS_DONE_BIT);
-        err_t err = dns_gethostbyname(aHostname, &addr, &wifi_dns_found_callback, &aResult);
-        if(err == ERR_OK && addr.u_addr.ip4.addr) {
-            aResult = addr.u_addr.ip4.addr;
-        } else if(err == ERR_INPROGRESS) {
-            waitStatusBits(WIFI_DNS_DONE_BIT, 15000);  //real internal timeout in lwip library is 14[s]
-            clearStatusBits(WIFI_DNS_DONE_BIT);
-        }
-        setStatusBits(WIFI_DNS_IDLE_BIT);
-        if((uint32_t)aResult == 0){
-            log_e("DNS Failed for %s", aHostname);
-        }
+    ip_addr_t addr;
+    aResult = static_cast<uint32_t>(0);
+    waitStatusBits(WIFI_DNS_IDLE_BIT, 16000);
+    clearStatusBits(WIFI_DNS_IDLE_BIT | WIFI_DNS_DONE_BIT);
+    err_t err = dns_gethostbyname(aHostname, &addr, &wifi_dns_found_callback, &aResult);
+    if(err == ERR_OK && addr.u_addr.ip4.addr) {
+        aResult = addr.u_addr.ip4.addr;
+    } else if(err == ERR_INPROGRESS) {
+        waitStatusBits(WIFI_DNS_DONE_BIT, 15000);  //real internal timeout in lwip library is 14[s]
+        clearStatusBits(WIFI_DNS_DONE_BIT);
+    }
+    setStatusBits(WIFI_DNS_IDLE_BIT);
+    if((uint32_t)aResult == 0){
+        log_e("DNS Failed for %s", aHostname);
     }
     return (uint32_t)aResult != 0;
 }
 
+/**
+ * Resolve the given hostname to an IP6 address.
+ * @param aHostname     Name to be resolved
+ * @param aResult       IPv6Address structure to store the returned IP address
+ * @return 1 if aHostname was successfully converted to an IP address,
+ *          else error code
+ */
+int WiFiGenericClass::hostByName6(const char* aHostname, ip_addr_t& aResult)
```

Comment:
I really think adding a separate hostByName6() method makes things worse, as it confuses a client user what they should call.

The should be able to just call hostByName(), pass in the DNS name, and get back the address (v6 or v4), whatever is supported by the current network -- and noting that networks, and hosts, can change over time, e.g. maybe today it has one IPv4 address, tomorrow it has a different one, in the future it is dual stack, and maybe one day it is IPv6 only.

The client shouldn't need to know or handle this. It just passes in the name, and gets back an IPAddress (which could be a v6 or v4).

A useful extension would, however, be hostsByName(), to get multiple addresses, in case part of the network is down... try address 1 (6 or 4), then address 2 (6 or 4), then address 3 (6 or 4), etc.

(The order of destination addresses should be as per RFC 6724 for default address selection.)

Reply:
> A useful extension would, however, be hostsByName(), to get multiple addresses, in case part of the network is down... try address 1 (6 or 4), then address 2 (6 or 4), then address 3 (6 or 4), etc.

That's rather tricky as it involves allocations and making the caller responsible to manage the de-allocation.
Also I don't think there are lots of use cases for IoT stuff where you will see multiple IP-addresses for the same hostname (apart from IPv4/IPv6)

Reply:
I don't see value in changing the APIs here. Multiple addresses should be managed by caller, and would be a rare case anyways.

Reply:
> That's rather tricky as it involves allocations and making the caller responsible to manage the de-allocation.

The way I've done it before is have a signature like `hostsByName(name, addresses[], max_results)`

And have the caller allocate an array for how many results they want, e.g. 4, so that you don't need any allocations. Not perfect, and there is a trade off between unused stack and dynamic allocation.

For destination addresses / DNS lookup, you probably are safe with just one of each (v6/v4), as usually a DNS lookup won't return scoped local addresses, ULAs, etc.

(In contrast `locaIP()` for IPv4 is usually only one address, either private, public, or link local, whereas for IPv6 there is usually at least two local IP addresses (link local and public), and possibly more (temporary public, other scopes, ULAs))

Reply:
At the very least, don't call the function hostByName6(), especially if it returns both v4 and v6 addresses.

My preferences/recommendation:

* Just have a single function, hostByName().
* With CONFIG_LWIP_IPV6 turned OFF, this functions exactly as before, and only returns v4 address
* With IPv6 turned on, it returns either a v6 or v4 (same as the suggested hostByName6()), with v6 as preference.
* This means to upgrade agnostic code to use IPv6, all you need to is turn the flag on. You never need to see the underlying IP addresses (or their type), just refer to destinations by name, e.g. "foo.example.com", and you will get the best address available (including if it changes in the future).
* Note that an address than only have v4, or on a v4 only network, will continue to use v4 so will still work fine. IPv6 is only used when available and supported.

Second option, if you want some support for diagnostic testing:

* Same as above, but when CONFIG_LWIP_IPV6 = ON, add a hostByName4() function, or a hostByName( ) function that takes a "type" parameter for v4/v6. This is normally no use, and no point to add it if V6 is off, but when V6 is on allows you to get a v4 response (if available) for debugging.
* I think a multiple address result (with the caller allocating) is more flexible, e.g. for diagnostics it allows multiple. But a secondary option is this alternative. (And documentation should mark it as only useful for diagnostics; in normal circumstances you don't need it, as hostByName() will return v4 when needed; filtering the result type is only useful when (a) you are running dual stack, (b) the network is dual stack, (c) the target is dual stack, but (d) you want to skip the IPv6 result and test the IPv4 result, e.g. for diagnostics. 

Third option (and a bad one):

* At least don't call it hostByName6(), which is misleading, but call it hostByName2() or hostByNameAny().
* This is a really messy upgrade path, as you can't just turn IPv6 only, but need to go around everywhere and change all calls to hostByName() to hostByName2().
* It also leaves a trap, that even if IPv6 is on and working fine, a coder may accidentally use the old hostByName() and get stuck in using legacy IPv4.
* All new code should be encourage to use the hostByName2() function
* The downside of this is that new code is then backwards compatible, i.e. if you have working dual stack code (using hostByName2()) but want to force IPv4 via CONFIG_LWIP_IPV6 = OFF (maybe your environment has a broken IPv6 implementation), then you need to also change all your code back to hostByName().
* This really is a bad option having an alternative function name.